### PR TITLE
Clusterctl integration

### PR DESCRIPTION
**Pull request by isaacdorfman**
**What this PR does / why we need it**:
In order to use capk with clusterctl, i.e: "clusterctl init", "clusterctl generate", the github repo must contain a release with manifests in the form specified by the clusterctl.
https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html

This PR is intended to create a github action that creates such release each time a PR is merged into the main branch.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release notes**:

```release-note
```

### CODE REVIEW

File: e2e/create-cluster_test.go [hunk 1]
```diff
@@ -402,6 +402,10 @@ var _ = Describe("CreateCluster", func() {
 			"--control-plane-machine-count=1",
 			"--worker-machine-count=1",
 			"--from", "templates/cluster-template.yaml")
+		cmd.Env = append(os.Environ(),
+			"NODE_VM_IMAGE_TEMPLATE=quay.io/capk/ubuntu-container-disk:20.04",
+			"IMAGE_REPO=k8s.gcr.io",
+			"CRI_PATH=/var/run/containerd/containerd.sock")
```
Comment (nirarg) [inline e2e/create-cluster_test.go:408]: This change is from Image builder PR, no?
https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/108

**Review by nirarg — COMMENTED** [review]
Its look like this PR include also changes from image builder PR (https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/108)
Please change it to include only relevant changes to this PR

**Comment by isaacdorfman** [comment]
@davidvossel 
I created a new makefile target called "functest-ubuntu" which uses environment variables that are relevant to the new image.
I need a way for the CI to invoke this target instead of the old one.

File: Makefile [hunk 2]
```diff
@@ -244,6 +243,17 @@ else
 	sed -i -e 's@imagePullPolicy: .*@imagePullPolicy: '"$(PULL_POLICY)"'@' ./config/default/manager_pull_policy.yaml
 endif
 
+## --------------------------------------
+## Deployment
+## --------------------------------------
+
+set_controller_image:
+	cd config/manager && kustomize edit set image controller=${CONTROLLER_IMG} > infrastructure-components.yaml
```
[the remark this answers is not in the document]
Comment (isaacdorfman) [reply Makefile:251]: I need to cd for kubestomize to work because the kustomization files are located there.

File: .github/workflows/create_release.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  build:
+    name: Upload Release Asset
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Set up Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: '1.16.6'
```
[the remark this answers is not in the document]
Comment (isaacdorfman) [reply .github/workflows/create_release.yaml]: Done

File: Makefile [hunk 4]
```diff
@@ -50,13 +50,10 @@ endif
 CONTROLLER_GEN := $(abspath $(TOOLS_BIN_DIR)/controller-gen)
 CONVERSION_GEN := $(abspath $(TOOLS_BIN_DIR)/conversion-gen)
 GOTESTSUM := $(abspath $(TOOLS_BIN_DIR)/gotestsum)
-KUSTOMIZE := $(abspath $(TOOLS_BIN_DIR)/kustomize)
-
-$(KUSTOMIZE): # Build kustomize from tools folder.
-	$(MAKE) -C $(ROOT) kustomize
+KUSTOMIZE ?= docker run k8s.gcr.io/kustomize/kustomize:v3.8.7
```
[the remark this answers is not in the document]
Comment (isaacdorfman) [reply Makefile]: Done

File: e2e/create-cluster_test.go [hunk 5]
```diff
@@ -428,6 +428,11 @@ var _ = Describe("CreateCluster", func() {
 			"--control-plane-machine-count=1",
 			"--worker-machine-count=1",
 			"--from", "templates/cluster-template.yaml")
+		cmd.Env = append(os.Environ(),
+			"NODE_VM_IMAGE_TEMPLATE=quay.io/capk/ubuntu-container-disk:20.04",
+			"IMAGE_REPO=k8s.gcr.io",
+			"CRI_PATH=/var/run/containerd/containerd.sock",
+		)
```
Comment (nirarg) [inline e2e/create-cluster_test.go:435]: Please remove this code

File: config/manager/kustomization.yaml [hunk 6]
```diff
@@ -1,2 +1,7 @@
 resources:
 - manager.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: controller
+  newName: localhost:5000/capk-manager
```
Comment (nirarg) [inline config/manager/kustomization.yaml:7]: Why is this needed?

File: e2e/create-cluster_test.go (hunk 5 above)
Comment (isaacdorfman) [reply e2e/create-cluster_test.go]: Done

**Review by nirarg — COMMENTED** [review]
You added two files named `infrastructure-components.yaml`, one under the root directory and one empty under `config/manager`
I guess this is mistake, please check

**Comment by davidvossel** [comment]
related to #101

File: infrastructure-components.yaml [hunk 7]
```diff
@@ -0,0 +1,5464 @@
...
+                                      type: object
+                                  required:
+                                  - spec
+                                  type: object
+                                type: array
+                              flavor:
+                                description: FlavorMatcher references a flavor that is used to fill fields in Template
+                                properties:
+                                  kind:
+                                    description: 'Kind specifies which flavor resource i
```
Comment (davidvossel) [inline infrastructure-components.yaml:5354]: this image needs to point to a specific image associated with the tag this infrastructure component yaml is generated from. Otherwise this image tag will continue to change in the future which will result in a mismatch of this yaml and the container image.

File: controllers/kubevirtcluster_controller.go [hunk 8]
```diff
@@ -84,8 +84,6 @@ func (r *KubevirtClusterReconciler) Reconcile(goctx gocontext.Context, req ctrl.
 		return ctrl.Result{}, nil
 	}
 
-	log = log.WithValues("cluster", cluster.Name)
```
Comment (davidvossel) [inline controllers/kubevirtcluster_controller.go:87]: why was this removed?

File: infrastructure-components.yaml [hunk 9]
```diff
@@ -0,0 +1,5464 @@
+apiVersion: v1
```
Comment (davidvossel) [inline infrastructure-components.yaml:1]: so we need to have this `infrastructure-components.yaml` in the source tree, or do we just need it uploaded as an asset to the release?

if we need it in the source tree, then we need something (a CI job) to ensure it's always up-to-date.  If we don't need this yaml for the quickstart in the source tree itself, then i say we don't commit it and only upload it in it's generated form to the github release.

**Review by davidvossel — COMMENTED** [review]
looks great, i just had some minor comments.

**Comment by davidvossel** [comment]
/hold

i'm trying to trigger e2e tests again.

File: controllers/kubevirtcluster_controller.go (hunk 8 above)
Comment (isaacdorfman) [reply controllers/kubevirtcluster_controller.go:87]: It causes the linter to fail, because this variable is declared and never used(I put this deletion in a separate commit)

File: infrastructure-components.yaml (hunk 9 above)
Comment (isaacdorfman) [reply infrastructure-components.yaml:1]: It only needs to be in the release.

File: controllers/kubevirtcluster_controller.go (hunk 8 above)
Comment (davidvossel) [reply controllers/kubevirtcluster_controller.go:87]: oh! alrighty

File: infrastructure-components.yaml (hunk 7 above)
Comment (isaacdorfman) [reply infrastructure-components.yaml]: Fixed

File: infrastructure-components.yaml (hunk 9 above)
Comment (isaacdorfman) [reply infrastructure-components.yaml]: Deleted and gitignored it.

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve
